### ISSUE TITLE
[Bug] Changing ListView CachingStrategy to "RecycleElement" stops updating cells when datasource updates in background

### ISSUE THREAD

**Issue by tonyfonager** [xamarin/Xamarin.Forms#14123]
### Description

In my app I have several ListViews which are bound to an ObservableCollection. This data updates frequently in the background, and new data streams into the app from the net, and when I update the data that the ListView is bound to, the ListView updates the cell with the changed data.

But - only if CachingStrategy is default ("RetainElement") - if I change the ListView to use CachingStrategy "RecycleElement", the cells stops getting the updates - that is 98% of the time. If I scroll up and down, sometimes the changed data is reflected inside the cells.

### Steps to Reproduce

1. Load the attached project, which is James Montemagnos "Monkeys" ListView demo app. I have added a timer to the ViewModel, so monkey names are updated every 1 seconds.
2. Run the project - you see that all Monkey names are updated every 1 second.
2. Now go into /Monkeys/Views/MokeysPage.xaml and change the ListView CachingStrategy, and set it to "RecycleElement" and rerun the sample.

### Expected Behavior

Data inside ListView should update whenever the data source changes, no matter what CachingStrategy is used.

### Actual Behavior

When using CachingStrategy "RetainElement" = correct behaviour and data is updated!

When using CachingStrategy "RecycleElement / RecycleElementAndDataTemplate" = bad bahaviour, and 99% of time data is not updated!

### Basic Information

- Version with issue: Tried both with XF 5.0 and demo project is XF 4.2
- Platform Target Frameworks: Issue see both on iOS and Android
- Affected Devices: All

### Reproduction Link

[Monkeys.zip](https://github.com/xamarin/Xamarin.Forms/files/6278305/Monkeys.zip)

Demo is based on a modified version of the ListView demo app "Monkeys" by James Montemagno : https://github.com/jamesmontemagno/Xamarin.Forms-Monkeys

**Comment by tonyfonager** [comment]
So absolutely nothing has happened in 19 days after I spent time creating and submitting this bug ?

Is Xamarin.Forms totally abandoned now that every XF dev wants to play with the new kid in town, Maui ?

**Comment by jamesmontemagno** [comment]
[App.zip](https://github.com/xamarin/Xamarin.Forms/files/6429319/App.zip)

I worked with Tony on Discord a bit.

The issue you have is that you are re-creating the entire list and who knows what is actually happening there
all sorts of bad potential

you want to implement INotifyPropertyChanged on the Monkey and raise it on the name

I don't actually think this is an issue with Xamarin.Forms and more with the code itself.

**Comment by jamesmontemagno** [comment]
@hartez ^

**Comment by hartez** [comment]
Argh, I literally just finished my reply to this and James beat me. 
I'm gonna post my reply anyway, since I took the time to write it :)

**Comment by hartez** [comment]
@tonyfonager 
In your repro project this is happening because the item you are updating doesn't implement `INotifyPropertyChanged`.

I'm attaching an updated version of the repro project - the only changes I've made are
- changing RetainElement to RecycleElement
- implementing INotifyPropertyChanged on the Monkey class :

```
public class Monkey : INotifyPropertyChanged
{
	private string name;

	public string Name
	{
		get => name; 
		set
		{
			name = value;
			OnPropertyChanged(nameof(Name));
		}
	}

	public event PropertyChangedEventHandler PropertyChanged;
	void OnPropertyChanged(string propertyName)
	{
		PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
	}

	/// other stuff omitted for brevity
}
```

In the background update method of your view model, you are replacing the entire bound items source. This means that the entire ListView is re-bound. With the `RetainElement` strategy, this means that a completely _new_ ViewCell is being created for every item in the source. At the time the ViewCells are created, they are bound to the current data, so they are (by definition) up-to-date. 

When you use the `RecycleElement` strategy, you aren't creating a brand new ViewCell for every single item; old ViewCells are re-used. When they are first created, these ViewCells are bound to a data item (an instance of the `Monkey` class, in your repro). You're then updating the `Name` property on each instance of `Monkey` and updating the `ItemSource`, which tells the ListView to re-bind everything. Since it's been told to recycle elements, it grabs the old ViewCell for each index and binds it to the appropriate instance of the `Monkey` class. In _most_ cases, it's the _exact same instance_ to which the ViewCell was already bound, so there's nothing for it to do. And the `Monkey` class doesn't have any mechanism for informing the DataBinding that any change has occurred, because it doesn't implement INotifyPropertyChanged. So the ViewCell doesn't update the Label, because nothing has told it to.

It's basically the same problem you'd see if you'd bound a single Label's `Text` property to the `Name` property of a `Monkey` instance and then updated `Name` - nothing would happen, because there's no notification mechanism. 

You noticed that some of the cells would update if you scrolled around - that's because when you scroll you're forcing it to create brand new ViewCells and/or making it re-use ViewCells that don't happen to already be bound to the data at their new index.

Your repro only _appears_ to work with `RetainElement` because it's completely recreating everything from scratch every time. 

tl;dr - RecycleElement isn't broken, you just need to implement INPC on your model so the ViewCells know that the Name has changed.

**Comment by hartez** [comment]
[MonkeysRecycleElement.zip](https://github.com/xamarin/Xamarin.Forms/files/6429411/MonkeysRecycleElement.zip)

**Comment by tonyfonager** [comment]
Thanks alot guys, I have read your detailed replies and will try implementing it inside my app!
